packages/backend-common/src/cache/CacheManager.ts

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2021 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Config, ConfigReader } from '@backstage/config';
+import cacheManager from 'cache-manager';
+// @ts-expect-error
+import Memcache from 'memcache-pp';
+// @ts-expect-error
+import memcachedStore from 'cache-manager-memcached-store';
+import { ConcreteCacheClient, CacheClient } from './CacheClient';
+import { PluginCacheManager } from './types';
+
+/**
+ * Implements a Cache Manager which will automatically create new cache clients
+ * for plugins when requested. All requested cacheclients are created with the
+ * credentials provided.
+ */
+export class CacheManager {
+  /**
+   * Keys represented supported `backend.cache.store` values, mapped to getters
+   * that return cacheManager.Cache instances appropriate to the store.
+   */
+  private readonly storeGetterMap = {
+    memcache: this.getMemcacheClient,
+    memory: this.getMemoryClient,
+    none: this.getNoneClient,
+  };
+
+  /**
+   * Creates a new CacheManager instance by reading from the `backend` config
+   * section, specifically the `.cache` key.
+   *
+   * @param config The loaded application configuration.
+   */
+  static fromConfig(config: Config): CacheManager {
+    // If no `backend.cache` config is provided, instantiate the CacheManager
+    // with empty config; allowing a "none" cache client will be returned.
+    return new CacheManager(
+      config.getOptionalConfig('backend.cache') || new ConfigReader(undefined),
+    );
+  }
+
+  private constructor(private readonly config: Config) {}
```

Comment:
Isn't it a bit unexpected to pass in the config to the constructor and parsing it over and over, when there is a `fromConfig`?

Could `fromConfig` do the parsing upfront and the constructor take a `(options: { ttl?: number }) => cacheManager.Cache` instead? We know the factory that's going to be used all the time anyway.

Reply:
Tried this out, and I like it for the `store` attribute, since that's fixed and should always be there.  But I ended up keeping a reference to the `backend.cache.connection` config because the configs will be quite different across stores and the built-in error throwing of the config getters is nice in the store-specific factory methods.